### ISSUE TITLE
Spawning Problems

### ISSUE THREAD

**Issue by C1ffisme** [PilzAdam/nether#5]
On the occasion, the portal will teleport you to the nether, and no room will spawn. This means you have to spend an hour flying up, or teleport.

* Solution 1: Ask paramat to make you a cavegen at nether depth, which will make sure you spawn in a cave, and will make a better nether.
* Solution 2: Try generating the room several times, just keep going until you detect a nether portal.
* Solution 3: Forget making a room. Use minetest.after(whatever_time_it_takes_to_generate_underground), and then search for a space to place the portal, place it, and teleport the player to the new portal position.

**Comment by sofar** [comment]
Is this still an issue?

**Comment by C1ffisme** [comment]
@sofar I'm not sure, I would have to download the mod myself and test it. Has the mapgen changed at all in all these years?

**Comment by Fixer-007** [comment]
It is still an issue :/ I've tried this on "Ultimate Diamond" server and was teleported to nether but no teleporter to go back.

**Comment by paramat** [comment]
> Solution 1: Ask paramat to make you a cavegen

Heheh.

**Comment by paramat** [comment]
```
					-- force emerge of target area
					minetest.get_voxel_manip():read_from_map(target, target)
					if not minetest.get_node_or_nil(target) then
						minetest.emerge_area(vector.subtract(target, 80), vector.add(target, 80))
					end
					-- teleport the player
					minetest.after(3, function(obj, pos, target)
```
The emerged volume is 3x3x3 = 27 mapchunks in size, this may take longer than 3 seconds to generate, perhaps try increasing that 3s or reducing the volume emerged.

**Comment by paramat** [comment]
Ugh you've got me thinking about this now, i will have to work on it.
It's possible that 'emerge_area' is not needed, as mapgen doesn't remove pre-existing nodes, this will speed things up a lot.

**Comment by sofar** [comment]
as said in the other issue - go ahead and make a PR.

**Comment by paramat** [comment]
Yes i'll do this properly now, i'll fork and make PRs.

**Comment by paramat** [comment]
Not sure if this has happened to me yet, possibly. I'm working on improving the creation of the destination portal, perhaps creating it not on teleport but earlier when the player-built portal is activated.

**Comment by paramat** [comment]
> On the occasion, the portal will teleport you to the nether, and no room will spawn

In testing this hasn't happened to me, however if the portal spawns in empty space the player will fall through it to whatever is below instead of standing in it. Maybe this was happening to you.

**Comment by Fixer-007** [comment]
No. Emerge area did not fix it. Experienced it on e748976. And yes, I've searched teleporter above and found nothing :/

**Comment by paramat** [comment]
Working on caverns i find that if the portal spawns in empty space the player will fall through it, as if it's generating after the mapgen and after player teleportation.

My plan is to create the destination portal on departure portal activation, not on player teleportation, so at least a few seconds earlier.

**Comment by paramat** [comment]
Tried it but on teleport to the nether the destination portal doesn't appear.

**Comment by C1ffisme** [comment]
@paramat What about a "minetest.register_on_mapgen_complete(function, params)?"

**Comment by paramat** [comment]
My new plan is to create the destination portal using a luavoxelmanip, this has the useful side effect of loading that volume into the client, which might help with teleportation issues.

**Comment by C1ffisme** [comment]
Ok.
